modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismLayoutInfo.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.sun.javafx.text;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import javafx.geometry.Insets;
+import javafx.geometry.Rectangle2D;
+import javafx.scene.text.CaretInfo;
+import javafx.scene.text.LayoutInfo;
+import javafx.scene.text.TextLineInfo;
+import com.sun.javafx.geom.BaseBounds;
+import com.sun.javafx.scene.text.TextLayout;
+import com.sun.javafx.scene.text.TextLine;
+
+/**
+ * Layout information as reported by PrismLayout.
+ */
+public abstract non-sealed class PrismLayoutInfo extends LayoutInfo {
+
+    protected abstract double lineSpacing();
+
+    protected abstract Insets insets();
+
+    private final TextLayout layout;
+
+    public PrismLayoutInfo(TextLayout layout) {
+        this.layout = layout;
+    }
+
+    @Override
+    public Rectangle2D getLogicalBounds(boolean includeLineSpacing) {
+        BaseBounds b = layout.getBounds();
+        Insets m = insets();
+        double dx = m.getLeft(); // TODO rtl?
+        double dy = m.getTop();
+        double sp = includeLineSpacing ? lineSpacing() : 0.0;
+        return TextUtils.toRectangle2D(b, dx, dy, sp);
+    }
+
+    @Override
+    public int getTextLineCount() {
+        return layout.getLines().length;
+    }
+
+    @Override
+    public List<TextLineInfo> getTextLines(boolean includeLineSpacing) {
+        TextLine[] lines = layout.getLines();
+        Insets m = insets();
+        double dx = m.getLeft(); // TODO rtl?
```

Comment:
This comment looks like it's a question to the reader. Does the code not work for RTL layouts?

Reply:
JavaFX does not support RTL fully - there is a whole series of issues waiting to be fixed.
See https://bugs.openjdk.org/browse/JDK-8343557